sumpy/p2p.py

```diff
@@ -52,76 +51,128 @@
 # {{{ p2p base class
 
 class P2PBase(KernelComputation, KernelCacheWrapper):
-    def __init__(self, ctx, kernels, exclude_self, strength_usage=None,
-            value_dtypes=None, name=None, device=None):
+    def __init__(self, ctx, out_kernels, exclude_self, strength_usage=None,
+            value_dtypes=None, name=None, device=None, in_kernels=None):
```

Comment:
I'm not sure I get why `in_kernels` and `out_kernels` makes sense for P2P. Could you explain?

Reply:
Let `ds` be some source derivative and `dt` be some target derivative.
To calculate  `H1 = ds(G) + G` and `H2 = dt(ds(G) + G)`, we need to calculate `ds(G), G` first which are the `in_kernels` and then the `out_kernels` would be `G, dt(G)` corresponding to `H1, H2`.

Reply:
Got it, thx.